### PR TITLE
修复 OpenCode 输出解析失败的BUG，并优化该CLI生成过程的反馈

### PR DESCRIPTION
**Pull request by ZhongXiandou**
解析 OpenCode 嵌套 part 输出，补充用量信息解析，减少重复 session 日志，并在等待生成时显示更准确的预览反馈。

### CODE REVIEW

**Comment by lefarcen** [comment]
嘿 @ZhongXiandou 👋

看了一下这个 PR 的范围 — 这块正在被多个 PR 同时改进中。简单交叉引用一下，避免重复工作：

**Related:**
- #52 (open) — 也在修 Codex JSON 解析
- #50 (open) — 同时支持 Codex/Gemini/Copilot CLI 的输出格式
- #26 (closed) — 之前描述了类似的 Codex JSON parsing drift 症状
- #27 (merged) — 上次 Codex 0.130 更新的修复

**Suggestion:** 你可能想和 #52/#50 的作者协调一下范围 — 这几个 PR 都在改同一个文件 (`argv.ts`)，可能存在冲突或者可以合并成一个更全面的修复。

审查进行中 👀

File: next/src/lib/agents/argv.ts [hunk 1]
```diff
@@ -362,7 +362,41 @@ function parseLineWithState(agent: string, line: string, state: ParseState): Age
...
+    const part =
+      obj.part && typeof obj.part === "object"
+        ? (obj.part as Record<string, unknown>)
+        : null;
+    if (typeof obj.text === "string") out.push({ kind: "delta", text: obj.text });
+    if (typeof obj.content === "string") out.push({ kind: "delta", text: obj.content });
+    if (typeof obj.message === "string") out.push({ kind: "delta", text: obj.message });
+    if (typeof part?.text === "string") out.push({ kind: "delta", text: part.text });
+    if (typeof part?.content === "string") out.push({ kind: "delta", text: part.content });
+    if (typeof part?.message === "string") out.push({ kind: "delta", text: part.message });
```
Comment (nettee) [inline next/src/lib/agents/argv.ts]: These new OpenCode lines append every matching alias field (`obj.text/content/message` and `part.text/content/message`). If one event carries the same chunk both at the top level and inside `part`—which is exactly the compatibility case this branch is trying to tolerate—`parseLine()` will emit two `delta` entries for one payload. `handleEvent()` then appends every delta directly to `task.html`, so the preview/output becomes duplicated or corrupted for affected runs. Please pick one preferred source per event (for example `part.text ?? part.content ?? part.message ?? obj.text ?? obj.content ?? obj.message`) or dedupe identical strings before pushing, and add a regression test that covers a payload containing both top-level and nested text fields.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

**Review by nettee — CHANGES_REQUESTED** [review]
I found one correctness issue in the new OpenCode parser path that can duplicate generated HTML for some event shapes.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

File: next/src/lib/agents/argv.ts (hunk 1 above)
Comment (ZhongXiandou) [reply next/src/lib/agents/argv.ts]: Thanks!

File: next/src/lib/agents/argv.ts [hunk 2]
```diff
@@ -362,7 +362,50 @@ function parseLineWithState(agent: string, line: string, state: ParseState): Age
...
+          ? part.content
+          : typeof part?.message === "string"
+            ? part.message
+            : typeof obj.text === "string"
+              ? obj.text
+              : typeof obj.content === "string"
+                ? obj.content
+                : typeof obj.message === "string"
+                  ? obj.message
+                  : null;
```
Comment (nettee) [inline next/src/lib/agents/argv.ts]: This precedence chain stops at the first *typed* alias, even when that alias is an empty string. For example, `parseLine("opencode", JSON.stringify({ type: "text", content: "<html>ok</html>", part: { type: "text", text: "" } }))` currently returns only `[{ kind: "delta", text: "" }]`. `handleEvent()` then drops that empty delta, so the valid top-level HTML chunk is lost entirely. Because this branch is meant to tolerate mixed top-level and nested aliases, it should choose the first **non-empty** string (or otherwise skip empty aliases before falling back) rather than the first string-valued field. Please update the selector to ignore `""` candidates and add a regression test for an empty nested alias plus a populated top-level alias.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

**Review by nettee — CHANGES_REQUESTED** [review]
I found one remaining correctness issue in the new OpenCode parser precedence logic.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

File: next/src/lib/agents/argv.ts (hunk 2 above)
Comment (ZhongXiandou) [reply next/src/lib/agents/argv.ts]: ok

File: next/src/lib/agents/argv.ts [hunk 3]
```diff
@@ -362,7 +362,39 @@ function parseLineWithState(agent: string, line: string, state: ParseState): Age
...
+        key: "usage",
+        value: {
+          input_tokens: tokens.input,
+          output_tokens: tokens.output,
+          cache_read_input_tokens: tokens.cache?.read,
+          cache_creation_input_tokens: tokens.cache?.write,
+        },
+      });
+    }
+    if (typeof part?.cost === "number") out.push({ kind: "meta", key: "cost_usd", value: part.cost });
```
Comment (nettee) [inline next/src/lib/agents/argv.ts]: `step_start` / `step_finish` are step-scoped events, but this branch forwards every `part.tokens` / `part.cost` payload as the final `usage` / `cost_usd` meta update. In `handleEvent()` those keys overwrite the task stats instead of accumulating them, so any OpenCode run with multiple steps will end up showing only the last step's tokens and cost. That makes the new usage feature report incorrect totals even though earlier steps consumed tokens too. Please either accumulate these step-finish values before emitting `usage` / `cost_usd`, or emit separate partial keys and sum them client-side before patching the run stats.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

**Review by nettee — COMMENTED** [review]
I found one follow-up issue in the new OpenCode usage parsing path.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

File: next/src/lib/agents/argv.ts (hunk 3 above)
Comment (ZhongXiandou) [reply next/src/lib/agents/argv.ts]: ok

**Review by nettee — APPROVED** [review]
@ZhongXiandou 我重新检查了这次 OpenCode 修复里的嵌套 part 文本提取、空字符串 fallback、多 step usage/cost 累加，以及等待态预览反馈这几处改动；当前 head 上这些问题链路已经对齐，新增的 argv 单测也覆盖了重复 delta、空嵌套字段回退和多步累加回归。这个 worktree 里未安装依赖，所以我没法在本地直接跑 Vitest，但从变更范围和现有测试覆盖看，这版实现是自洽的。辛苦了，这轮跟进修得很扎实。


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=opencode · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by ZhongXiandou** [comment]
还不合并吗🤣

**Comment by lefarcen** [comment]
看到了 @ZhongXiandou 👀

这边已经同步值班 maintainer 做合并判断了。当前 PR 已有 @nettee approve，GitHub 也显示可合并；因为这个仓库没有 CI 信号，最后合并还需要 maintainer 手动确认一下。

**Comment by lefarcen** [comment]
补充同步一下：#67 里 reporter 已确认同一类 OpenCode 复现路径（OpenCode `exit=0`，但预览区仍停在模板 HTML），这个现象和本 PR 修的嵌套 OpenCode 输出解析方向是对得上的。

我刚重新看了 PR 当前公开状态：@nettee 已 approve，Public checks 是 green，merge state 也是 clean/mergeable。已经把这个更新状态同步给 maintainer 做最终合并判断了。

**Comment by lefarcen** [comment]
Heads-up: #92 is now also open against this area — both PRs touch `next/src/lib/agents/argv.ts` and address the OpenCode output parsing path tied to #67. #92 is a narrower fix for the current singular `part.text` JSON-line shape, while this PR includes broader progress/usage handling too. @ZhongXiandou and @Aliancn may want to compare approaches; the maintainer team will decide which path lands.
